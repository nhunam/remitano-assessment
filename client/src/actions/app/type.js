import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  CLEAR_MESSAGE: undefined
});
