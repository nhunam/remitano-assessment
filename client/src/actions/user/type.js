import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  USER_CREATE: undefined,
  USER_LOGIN: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined
});
