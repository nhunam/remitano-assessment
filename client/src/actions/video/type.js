import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  VIDEO_FETCH: undefined,
  VIDEO_FETCH_SUCCESS: undefined,
  VIDEO_FETCH_FAILURE: undefined,
  VIDEO_SHARING: undefined,
  VIDEO_SHARING_SUCCESS: undefined,
  VIDEO_SHARING_FAILURE: undefined
});
