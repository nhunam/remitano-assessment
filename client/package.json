{
  "name": "video-sharing-react",
  "version": "0.1.0",
  "private": true,
  "jest": {
    "collectCoverageFrom": [
      "src/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/node_modules/",
      "!<rootDir>/src/containers/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/actions/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/configurations/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/App.jsx",
      "!<rootDir>/src/serviceWorker.js",
      "!<rootDir>/src/index.js"
    ],
    "coverageThreshold": {
      "global": {
        "branches": 90,
        "functions": 90,
        "lines": 90,
        "statements": 90
      }
    }
  },
  "dependencies": {
    "axios": "^0.19.0",
    "bootstrap": "^4.4.1",
    "classnames": "^2.2.6",
    "enzyme": "^3.10.0",
    "enzyme-adapter-react-16": "^1.15.1",
    "flow-bin": "^0.113.0",
    "http-proxy-middleware": "^1.0.6",
    "immutability-helper": "^3.0.1",
    "jwt-decode": "^2.2.0",
    "keymirror": "^0.1.1",
    "node-sass": "^4.13.0",
    "react": "^16.12.0",
    "react-dom": "^16.12.0",
    "react-infinite-scroller": "^1.2.4",
    "react-player": "^1.14.2",
    "react-redux": "^7.1.3",
    "react-router": "^5.1.2",
    "react-router-dom": "^5.1.2",
    "react-scripts": "3.2.0",
    "react-test-renderer": "^16.12.0",
    "redux": "^4.0.4",
    "redux-actions": "^2.6.5",
    "redux-logger": "^3.0.6",
    "redux-mock-store": "^1.5.3",
    "redux-persist": "^6.0.0",
    "redux-saga": "^1.1.3",
    "serialize-javascript": "2.1.1",
    "styled-components": "^4.4.1"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "flow": "flow"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "redux-saga-test-plan": "^4.0.0-rc.3"
  }
}
