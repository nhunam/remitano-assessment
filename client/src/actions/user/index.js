import { ActionTypes } from "./type";
import { createActions } from "redux-actions";

export const {
  userLogin: login,
  userLogout: logout,
  userCreate: create
} = createActions({
  [ActionTypes.USER_LOGIN]: param => ({ ...param }),
  [ActionTypes.USER_CREATE]: param => ({ ...param }),
  [ActionTypes.USER_LOGOUT]: () => ({})
});
