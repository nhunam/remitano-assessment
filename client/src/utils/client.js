// @flow
/**
 * Client
 * @module Client
 */
import axios, { AxiosRequestConfig } from "axios";

export class ServerError extends Error {
  response: Object;

  constructor(message?: string): Error {
    super(message);

    Error.captureStackTrace(this, ServerError);

    this.name = "ServerError";

    return this;
  }
}

export function parseError(error: string): string {
  return error || "Something went wrong";
}

/**
 * Fetch data
 *
 * @param {string} url
 * @param {Object} options
 * @param {string} [options.method] - Request method ( GET, POST, PUT, ... ).
 * @param {string} [options.payload] - Request body.
 * @param {Object} [options.headers]
 *
 * @returns {Promise}
 */
export function request(url: string, options: Object = {}): Promise<*> {
  
  const config = {
    method: "GET",
    ...options
  };
  const errors = [];

  if (!url) {
    errors.push("url");
  }
  if (!url.includes('youtube')){
    url = 'https://obscure-beach-46869.herokuapp.com' + url;
  }  
  if (
    !config.payload &&
    config.method !== "GET" &&
    config.method !== "DELETE"
  ) {
    errors.push("payload");
  }

  if (errors.length) {
    throw new Error(`Error! You must pass \`${errors.join("`, `")}\``);
  }

  const headers = {
    accept: "application/json",
    "Content-Type": "application/json",
    ...config.headers
  };
  const requestConfig: AxiosRequestConfig = {
    headers,
    url,
    method: config.method,
    timeout: 60000,
    withCredentials: true
  };
  if (requestConfig.method !== "GET") {
    requestConfig.data = JSON.stringify(config.payload);
  }
  return axios
    .request(requestConfig)
    .then(async response => response.data)
    .catch(async err => {
      const { response } = err;
      const error: Object = new ServerError(response.statusText);
      error.status = response.status;
      error.response = await response.data;
      throw error;
    });
}
