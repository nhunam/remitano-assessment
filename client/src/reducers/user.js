import { ActionTypes } from "actions/user/type";
import { AppStatus } from "configurations";
import { handleActions } from "redux-actions";
import immutable from "immutability-helper";

export const userState = {
  isAuthenticated: false,
  info: {
    id: -1,
    email: "",
    token: ""
  }
};

export default {
  user: handleActions(
    {
      [ActionTypes.USER_LOGIN_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          isAuthenticated: { $set: true },
          info: {
            id: { $set: payload.id },
            email: { $set: payload.email },
            token: { $set: payload.token }
          }
        }),
      [ActionTypes.USER_LOGOUT]: state =>
        immutable(state, {
          status: { $set: AppStatus.RUNNING }
        }),
      [ActionTypes.USER_LOGOUT_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: false },
          info: { id: { $set: -1 }, email: { $set: "" }, token: { $set: "" } },
          status: { $set: AppStatus.IDLE }
        })
    },
    userState
  )
};
