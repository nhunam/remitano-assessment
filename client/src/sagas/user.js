import { all, call, put, takeLatest } from "redux-saga/effects";

import { ActionTypes } from "actions/user/type";
import jwt_decode from "jwt-decode";
import { request } from "utils/client";

/**
 * Login
 */
export function* authenticate(param) {
  try {
    const payload = { method: "POST", ...param };
    const response = yield call(request, "/token", payload);
    const { token } = response;
    const user = jwt_decode(token);

    yield put({
      type: ActionTypes.USER_LOGIN_SUCCESS,
      payload: { token: token, email: user.email, id: user.id }
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.USER_LOGIN_FAILURE,
      payload: err
    });
  }
}

/**
 * create
 */
export function* create(param) {
  const { email, password } = param.payload;
  try {
    const payload = { method: "POST", ...param };
    yield call(request, "/users", payload);

    yield put({
      type: ActionTypes.USER_LOGIN,
      payload: { email: email, password: password }
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.USER_LOGIN,
      payload: { email: email, password: password }
    });
  }
}

/**
 * Logout
 */
export function* logout() {
  yield put({
    type: ActionTypes.USER_LOGOUT_SUCCESS
  });
}

/**
 * User Sagas
 */
export default function* root() {
  yield all([
    takeLatest(ActionTypes.USER_LOGIN, authenticate),
    takeLatest(ActionTypes.USER_LOGOUT, logout),
    takeLatest(ActionTypes.USER_CREATE, create)
  ]);
}
