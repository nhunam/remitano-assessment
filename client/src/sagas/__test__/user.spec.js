import * as matchers from "redux-saga-test-plan/matchers";

import user, { authenticate, create, logout } from "sagas/user";

import { ActionTypes } from "actions/user/type";
import { expectSaga } from "redux-saga-test-plan";
import { request } from "utils/client";

describe("User Sagas", () => {
  it("should have the expected watchers", done => {
    expectSaga(user)
      .run({ silenceTimeout: true })
      .then(saga => {
        expect(saga).toMatchSnapshot();
        done();
      });
  });

  it("should match the authenticate saga", () => {
    const fakeToken = {
      type: "bearer",
      token:
        "eyJhbGciOiJIUzUxMiJ9.eyJpZCI6MSwiZXhwIjoxNTc1NTQ4Nzc3LCJpYXQiOjE1NzU1MzA3NzcsImVtYWlsIjoidGVzdDJAZ21haWwuY29tIn0.RClf9eFWY6hKoKqhm6RWY8PCXlAEj6eNNk8YOtbbZMzjF_vpp_wZc3a2aUddQYXHTMJ-1jO3_Agf76Zani9ETg"
    };
    expectSaga(authenticate)
      .provide([[matchers.call.fn(request), fakeToken]])
      .put({
        type: ActionTypes.USER_LOGIN_SUCCESS,
        payload: {
          token:
            "eyJhbGciOiJIUzUxMiJ9.eyJpZCI6MSwiZXhwIjoxNTc1NTQ4Nzc3LCJpYXQiOjE1NzU1MzA3NzcsImVtYWlsIjoidGVzdDJAZ21haWwuY29tIn0.RClf9eFWY6hKoKqhm6RWY8PCXlAEj6eNNk8YOtbbZMzjF_vpp_wZc3a2aUddQYXHTMJ-1jO3_Agf76Zani9ETg",
          email: "test2@gmail.com",
          id: 1
        }
      })
      .run();
  });

  it("should match the create saga", () => {
    const fakeUser = {
      id: 1,
      email: "test@gmail"
    };
    expectSaga(create, { payload: { email: "email", password: "password" } })
      .provide([[matchers.call.fn(request), fakeUser]])
      .put({
        type: ActionTypes.USER_LOGIN,
        payload: { email: "email", password: "password" }
      })
      .run();
  });

  it("should match the logout saga", () =>
    expectSaga(logout)
      .put({
        type: ActionTypes.USER_LOGOUT_SUCCESS
      })
      .run());
});
