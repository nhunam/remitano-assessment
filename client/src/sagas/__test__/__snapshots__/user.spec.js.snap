// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Sagas should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "USER_LOGIN",
          "@@redux-saga-test-plan/json/function/authenticate",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "USER_LOGOUT",
          "@@redux-saga-test-plan/json/function/logout",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "USER_CREATE",
          "@@redux-saga-test-plan/json/function/create",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "USER_LOGIN",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "USER_LOGOUT",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "USER_CREATE",
      },
      "type": "TAKE",
    },
  ],
}
`;
